### PR TITLE
Heartbeat

### PR DESCRIPTION
**Pull request by bonnieWeng**
The current program compiles. Commented out line 18 in heartbeat_test.c, which causes a compiler error saying: undefined reference to... 

The heartbeat lib common format (version 3) is not entirely complete (but mostly there!). Would like some feedback and help with the bug.

### CODE REVIEW

File: src/heartbeat/heartbeat.c [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    switch(ssm_id){
+        case 0b00:
+            self_status = &obc_status;
+            parent_status = &eps_status;
+            child_status = &pay_status;
+            break;
+        case 0b10:
+            self_status = &eps_status;
+            parent_status = &pay_status;
+            child_status = &obc_status;
```
Comment (siddMahen) [inline src/heartbeat/heartbeat.c]: Missing break statements here.

File: src/heartbeat/heartbeat.c [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        case 0b01:
+            self_status = &pay_status;
+            parent_status = &obc_status;
+            child_status = &eps_status;
```
Comment (siddMahen) [inline src/heartbeat/heartbeat.c]: And here.

**Comment by siddMahen** [comment]
@bonnieweng1T9 I was thinking about EEPROM reliability and max read/write count (documentation says 10^5 write-erase cycles); Do you think it would be useful to add a counter somewhere in EEPROM that measures how many total writes occur. We could do this by wrapping the avr eeprom functions such that they increment some counter, also in EEPROM.

**Comment by bonnieweng1T9** [comment]
Hey @siddMahen ! I love the idea!! I think it would really useful as our current software program does rely on EEPROM, and having an index of EEPROM reliability is essential to our program. Could you please elaborate on your idea on how exactly you wanna do this? What do you mean by wrapping EEPROM functions? or if you don't mind, we can talk tmr briefly at the general meeting as well!

**Comment by bonnieweng1T9** [comment]
@siddMahen Help plz!! Heartbeat lib common is complete in this commit. However, I commented out some codes from line 53-67 (under the assign_CAN_status_mobs function). I was trying to modify the id_tags using switch statements based on SSM type, but it gives compilation errors and I dunno how to fix. Plz bless your computing power. Testing to come very soon! And if any feedback, I'm all to hear.

**Comment by siddMahen** [comment]
This branch **cannot** be merged until it's git history is cleaned up, using `git rebase`. @bonnieweng1T9 when it's almost done, I'll clean it up

**Comment by brunofalmeida** [comment]
I recommend defining the SSM IDs (0x00, 0x01, 0x02) using `#define` constants for OBC, PAY, and EPS boards so it's clear.

**Comment by bonnieweng1T9** [comment]
Like your idea, @brunofalmeida Will clean it up this evening/ tonight!

File: src/heartbeat/heartbeat.c [hunk 3]
```diff
@@ -1,37 +1,83 @@
-//Assume init_uart() and init_can() have been called
+/*
```
Comment (siddMahen) [inline src/heartbeat/heartbeat.c]: We can leave this in for now, but in the future I plan on removing all of these headers.

I find that the format doesn't supply much useful information; the filename and dependencies are obvious; the authors, date modified and revision history are all available via git. The notes and description section are useful, but feel out of place.

Also, I find that people tend to litter this section with tabs and spaces, and there are tons of alignments issues, which looks awful.

**Comment by siddMahen** [comment]
@bonnieWeng is this completely done? If so, I'll rebase and merge.

**Comment by bonnieWeng** [comment]
The current version of heartbeat is ready, but @BrytniR and I oversee some **integration challenge**. As a quick overview, this is what the current heartbeat does: 

1. **Retrieve the latest status after a SSM restart**, by calling `init_heartbeat()`, which retrieves the latest status from EEPROM and store status data into status global variables. 
2. **Keep track of status data among each 3 SSM**, using the triangular SSM configuration as described in heartbeat.c. The current set up is, when there is a status change in SSM main program, the program will need to call `heartbeat()`, which sends a tx to its parent to inform about the updated status change.

Current Challenges

1.  **can ids**: The current can set up does not follow the universal rx masking protocol as outlined in can_ids.h. According to @vogeldylan, this is something we should fix.
2. **What are the status data?** And how does it fit into the current command queue structure (OBC) or command handling structure (EPS & PAY)?
3. **The current code does not guarantee that all 3 SSMs will have identical status data at any given snapshot of time** And here's the reason why: We design heartbeat such that when there is a status change, a SSM needs to inform its parent about such change (via STATUS TX). A status change is defined as any of the OBC, EPS or PAY status change. By that design, it implies that, when a SSM receives STATUS RX, there has been a status change, and that SSM would need to send STATUS TX to its parent, however, **this is currently NOT implemented**. The current heartbeat set up is to have a SSM sends STATUS TX to its parent when its own status changes in its SSM main program.

Design Re-consideration

One thing important to understand heartbeat, the primary goal is **status recovery at SSM restart**. By this rationale, sending STATUS message among each SSM is redundant, however, there might be benefits to it, if we utilize the design appropriately. And this back to the initial design question, what else can we benefit from the decentralized authority of SSM management?

File: src/heartbeat/heartbeat.c (hunk 3 above)
Comment (bonnieWeng) [reply src/heartbeat/heartbeat.c]: 1. Agree filename, dependencies and revision history are redundant. And agree notes and description are out of place (I do feel that too when I edited). 
2. I actually copy the format from PAY code. They have been using this format for quite awhile. Maybe, if you would like, we can set up an universal way on how to document descriptions and notes on our code base?! @vogeldylan @brunofalmeida

**Comment by siddMahen** [comment]
@bonnieWeng I think the content of the statuses and the CAN id issues can be dealt with relatively easily, they shouldn't be very problematic when it comes to integration. You, @vogeldylan @alihaydaroglu, and I should meet to discuss the status contents.

Regarding the third point, I guess that's still TBD. I feel as though the additional guarantees might be worth it.

File: src/heartbeat/heartbeat.c (hunk 3 above)
Comment (brunofalmeida) [reply src/heartbeat/heartbeat.c]: Actually, I was working on cleaning this up in the pay qual-development branch yesterday. I agree that filename, dependencies, and revision history are redundant.

Comment (brunofalmeida) [reply src/heartbeat/heartbeat.c]: And authors are probably redundant too

Comment (bonnieWeng) [reply src/heartbeat/heartbeat.c]: I would keep the authors. :p Cuz people who contribute to the design, might not be the person committing changes. The purpose of having authors is when there's a question in codes, you know who to ask. And for UTAT, there has been trouble asking people about past work cuz people come and go. So more reference, the safer in my opinions.

Comment (brunofalmeida) [reply src/heartbeat/heartbeat.c]: Alright, that makes sense

**Pull request merged**